### PR TITLE
Optional repo cache for offline faster compile

### PR DESCRIPTION
**Pull request by marbx**
### What does this PR do?
Optionally keep/cache required software,
The user must set two enviromental variables to activate the cache.
Without them, current behaviour is unchanged.

### What issues does this PR fix?
My company proxy refuses connection for an admin user.
Therefore, I cannot compile Salt within my company.
As an additional benefit, compile time goes down from 3:30 to 2:10.
After 240 compiles, I will actually save time...

### Previous Behavior
Every required software was downloaded for each compile.

### New Behavior
Two directories cache saltrepo and pypy, 130 MB in 50 files are kept local.
Some downloads remain, and will be dealt with soon.

### Tests written?
Yes, tested both cases (with and without the environment variables).

### CODE REVIEW

File: pkg/windows/build_env.ps1 [hunk 1]
```diff
@@ -87,10 +91,17 @@ If (!(Get-IsAdministrator)) {
...
 
+
 #------------------------------------------------------------------------------
-# Create Directories
+# You can set SALTREPO_LOCAL_CACHE and SALTREPO_LOCAL_CACHE_PIP to  permanent cache directories (DownloadDir). 
+# The script empties DownloadDir unless SALTREPO_LOCAL_CACHE environment variable set
 #------------------------------------------------------------------------------
-$p = New-Item $ini['Settings']['DownloadDir'] -ItemType Directory -Force
+if ( ! [bool]$Env:SALTREPO_LOCAL_CACHE ) {
+  $p = New-Item $ini['Settings']['DownloadDir'] -ItemType Directory -Force
```
Comment (twangboy) [inline pkg/windows/build_env.ps1]: Please use 4 spaces per indentation level.

File: pkg/windows/build_env.ps1 [hunk 2]
```diff
@@ -206,73 +216,83 @@ If (!($Path.ToLower().Contains("$($ini['Settings']['ScriptsDir'])".ToLower())))
...
 #==============================================================================
 Write-Output " ----------------------------------------------------------------"
-Write-Output " - Installing pypi resources using pip . . ."
+Write-Output " - $script_name :: Installing pypi resources using pip . . ."
 Write-Output " ----------------------------------------------------------------"
-Start_Process_and_test_exitcode "$($ini['Settings']['ScriptsDir'])\pip.exe"  "--no-cache-dir install -r $($script_path)\req.txt" "pip install"
+if ( ! [bool]$Env:SALTREPO_LOCAL_CACHE_PIP) {
+  Start_Process_and_test_exitcode "$($ini['Settings']['ScriptsDir'])\pip.exe"  "--no-cache-dir install -r $($script_path)\req.txt" "pip install"
+} else {
+  if ( (Get-ChildItem $Env:SALTREPO_LOCAL_CACHE_PIP | Measure-Object).Count -eq 0 ) {
```
Comment (twangboy) [inline pkg/windows/build_env.ps1]: Again here... 4 spaces per indentation level

File: pkg/windows/build_env.ps1 [hunk 3]
```diff
@@ -286,7 +306,9 @@ If (-Not $Silent) {
 # Remove the temporary download directory
 #------------------------------------------------------------------------------
 Write-Output " ----------------------------------------------------------------"
-Write-Output " - Cleaning up downloaded files"
+Write-Output " - $script_name :: Cleaning up downloaded files unless you use SALTREPO_LOCAL_CACHE"
 Write-Output " ----------------------------------------------------------------"
 Write-Output ""
+if ( ! [bool]$Env:SALTREPO_LOCAL_CACHE ) {
 Remove-Item $($ini['Settings']['DownloadDir']) -Force -Recurse
```
Comment (twangboy) [inline pkg/windows/build_env.ps1]: Please indent this line as it is now in an `if` block

File: pkg/windows/modules/download-module.psm1 [hunk 4]
```diff
@@ -16,43 +22,55 @@ Function DownloadFileWithProgress {
     )
 
     begin {
-        $client = New-Object System.Net.WebClient
-        $Global:downloadComplete = $false
-        $eventDataComplete = Register-ObjectEvent $client DownloadFileCompleted `
+        $Global:NEED_PROCESS_AND_END = $true
+        # Write-Host " **** DownloadFileWithProgress looking for **** $localFile ********"
+        if ( [bool]$Env:SALTREPO_LOCAL_CACHE -and (Test-Path $localFile) ) { 
+          # Write-Host " **** found **** $localFile ********"
```
Comment (twangboy) [inline pkg/windows/modules/download-module.psm1]: again Indentation level...

File: pkg/windows/modules/download-module.psm1 [hunk 5]
```diff
@@ -16,43 +22,55 @@ Function DownloadFileWithProgress {
...
+          $eventDataProgress = Register-ObjectEvent $client DownloadProgressChanged `
             -SourceIdentifier WebClient.DownloadProgressChanged `
             -Action { $Global:DPCEventArgs = $EventArgs }
+        }
     }
     process {
-        Write-Progress -Activity 'Downloading file' -Status $url
-        $client.DownloadFileAsync($url, $localFile)
+        if ( $Global:NEED_PROCESS_AND_END ) { 
+          #Write-Host " ++++++ actually DOWNLOADING ++++++ $localFile +++++++"
```
Comment (twangboy) [inline pkg/windows/modules/download-module.psm1]: And here....

File: pkg/windows/modules/download-module.psm1 [hunk 6]
```diff
@@ -16,43 +22,55 @@ Function DownloadFileWithProgress {
...
-        $Global:DPCEventArgs = $null
-        Remove-Variable client
-        Remove-Variable eventDataComplete
-        Remove-Variable eventDataProgress
-        [GC]::Collect()
-        # 2016-07-06  mkr  Errorchecking added. nice-to-have: integration into the above code.
-        If (!((Test-Path "$localfile") -and ((Get-Item "$localfile").length -gt 0kb))) {
-            Write-Error "Exiting because download missing or zero-length:    $localfile"
+        if ( $Global:NEED_PROCESS_AND_END ) { 
+          Unregister-Event -SourceIdentifier WebClient.DownloadProgressChanged
```
Comment (twangboy) [inline pkg/windows/modules/download-module.psm1]: Again here...

File: pkg/windows/modules/get-settings.psm1 [hunk 7]
```diff
@@ -20,7 +20,14 @@ Function Get-Settings {
             "ScriptsDir"  = "C:\Python27\Scripts"
             "DownloadDir" = "$env:Temp\DevSalt"
             }
+        # The script deletes the DownLoadDir (above) for each install.
+        # You may want to set an environment variable SALTREPO_LOCAL_CACHE, a cache which lives as long as you decide.
+        if ( [bool]$Env:SALTREPO_LOCAL_CACHE ) {
+          $Settings.Set_Item("DownloadDir", "$Env:SALTREPO_LOCAL_CACHE")
```
Comment (twangboy) [inline pkg/windows/modules/get-settings.psm1]: And here...

**Review by twangboy — COMMENTED** [review]
Just a few formatting issues. 4 spaces per indent level.

**Comment by markuskramerIgitt** [comment]
Hi @twangboy,
with https://github.com/markuskramerIgitt/salt/commit/acb54b3da7e10555cca7012277adb1fe8e926222
I corrected indentation in *.ps1 and *.psm1 by replacing withespace with tabs.

The 7 files files above went into this branch (`Optional_repo_cache_for_offline_faster_compile`), 
the other 2 (download_url_file.ps1 and build_env.ps1) will shortly go into branch `replace_bitsadmin`.

I still need to cache these 2:
 - pip==8.1.2 (from -r C:\git\salt\pkg\windows\req_pip.txt (line 1))
 - setuptools==25.1.3 (from -r C:\git\salt\pkg\windows\req_pip.txt (line 2))

**Comment by markuskramerIgitt** [comment]
Hello @twangboy, since 1 hour I cannot figure out how to create a pull request with just these two files in in
 - download_url_file.ps1
 - build_env.ps1
I created a branch `replace_bitsadmin`, but I always have 7 files in the pull request.

Can you tell me, or shall I ask Nicole?

Best regards,
Markus

**Comment by markuskramerIgitt** [comment]
Hi @twangboy,
I figured it out how to create pull request https://github.com/saltstack/salt/pull/38924 of only the above two files. It replaces bitsadmin with DownloadFileWithProgress.

---
I still need to cache these 2 resources:
 - pip==8.1.2 (from -r C:\git\salt\pkg\windows\req_pip.txt (line 1))
 - setuptools==25.1.3 (from -r C:\git\salt\pkg\windows\req_pip.txt (line 2))

**Comment by twangboy** [comment]
@markuskramerIgitt They need to be spaces, not tabs. You had it right the first time.

**Comment by markuskramerIgitt** [comment]
Hi @twangboy,
I am aware that Python and sls files need 4 spaces, because tabs could lead to ambiguity.

Powershell does not suffer from this ambiguity, and I would like to use tabs and represent indentation  according your personal taste.

Can I kindly ask for this flexibility?

**Comment by cachedout** [comment]
Bump @twangboy

**Comment by markuskramerIgitt** [comment]
Hi @twangboy, 
I made git mistakes, all my pull requests contain the same 10 files. 

I will now cancel my pull requests and issue new ones.

**Pull request closed without merging**